### PR TITLE
Issue #3546: add lambda in a rightcurlycheck

### PR DESCRIPTION
**Pull request by Vladlis**
Issue #3546

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 1]
```diff
@@ -320,19 +346,9 @@ private static Details getDetails(DetailAST ast) {
 
         switch (ast.getType()) {
             case TokenTypes.LITERAL_TRY:
-                if (ast.getFirstChild().getType() == TokenTypes.RESOURCE_SPECIFICATION) {
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Why are there changes to `try`? I'm expecting only changes for lambdas.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 2]
```diff
@@ -445,6 +468,21 @@ private static DetailAST getNextToken(DetailAST ast) {
     }
 
     /**
+     * Checks that before a right curly should be a linebreak.
+     * @param bracePolicy options for placing the right curly brace.
+     * @param details Details for validation
+     * @return true if before a right curly should be a linebreak.
+     */
+    private static boolean shouldHaveBeLineBreakBefore(RightCurlyOption bracePolicy,
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: `shouldHaveBeLineBreakBefore` -> `shouldHaveLineBreakBefore`

File: src/main/resources/google_checks.xml [hunk 3]
```diff
@@ -61,7 +61,7 @@
         <module name="RightCurly">
             <property name="id" value="RightCurlyAlone"/>
             <property name="option" value="alone"/>
-            <property name="tokens" value="CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE, STATIC_INIT, INSTANCE_INIT"/>
+            <property name="tokens" value="CLASS_DEF, METHOD_DEF, CTOR_DEF, LITERAL_FOR, LITERAL_WHILE, STATIC_INIT, INSTANCE_INIT, LAMBDA"/>
```
Comment (rnveach) [inline src/main/resources/google_checks.xml]: We need confirmation on google style guides requirements for this from http://checkstyle.sourceforge.net/reports/google-java-style-20160712.html or https://google.github.io/styleguide/javaguide.html .

**Comment by rnveach** [comment]
This PR will have to wait until PR #3895 is done and be rebased off that.

**Comment by romani** [comment]
@Vladlis , please do not wait.
PR that is completed first is merged first, all other will need to rebase.

File: src/main/resources/google_checks.xml (hunk 3 above)
Comment (romani) [inline src/main/resources/google_checks.xml]: https://google.github.io/styleguide/javaguide.html#s4.1.2-blocks-k-r-style
Looks like is clear by example on rightfully location.

@rnveach , please clarify your request.

Comment (rnveach) [inline src/main/resources/google_checks.xml]: Sorry, I forgot I also wrote that in the issue. that is all I was looking for was the section that specified it.

**Comment by MEZk** [comment]
@Vladlis 
False negative:

```java
void foo1() {
        final java.util.function.Predicate<String> isPolindrome = new java.util.function.Predicate<String>() {
            @Override
            public boolean test(String s) {
                final String reversed =
                    new StringBuilder(s.replaceAll("\\W", "")).reverse().toString();
                return reversed.equalsIgnoreCase(s);
            }
        };

        final java.util.List<String> polindromes =
            java.util.stream.Stream.of("Hello", "World", "flee to me remote elf", null, "")
            .filter(s -> {
                return s != null
                    && !s.trim().isEmpty()
                    && isPolindrome.test(s);
            }) // here should be a violation
            .collect(java.util.stream.Collectors.toList());
    }
```

**Comment by Vladlis** [comment]
@romani 
how do we resolve failures in non-error tests? do I need to provide a PR to hibernate-search?

@MEZk 
thanks

**Comment by romani** [comment]
>do I need to provide a PR to hibernate-search?

Yes. If they do not merge your PR quickly (in 3-4days) , we do fork of them, merge your fix to our fork, switch CI to clone from fork.

**Comment by romani** [comment]
@Vladlis , please rebase your changes.

**Comment by rnveach** [comment]
@Vladlis ping

**Comment by Vladlis** [comment]
The check is broken in the current state. I guess it should be fixed before I proceed with a new functionality.
During this PR I've already found https://github.com/checkstyle/checkstyle/issues/4044 and now: https://github.com/checkstyle/checkstyle/issues/4085.

**Comment by romani** [comment]
issue #4085 is approved.

**Comment by Vladlis** [comment]
@romani , it I submit PR now it will introduce a bunch of false negatives

**Comment by Vladlis** [comment]
@romani, one more issue: https://github.com/checkstyle/checkstyle/issues/4091

**Comment by MEZk** [comment]
@Vladlis 
Is this PR postponed till #4091 and ... ?

**Comment by Vladlis** [comment]
@MEZk , yes, right now #4091 is the only blocker I see

**Comment by romani** [comment]
@Vladlis , please rebase.
Blocker issue will be reviewed asap.

**Comment by Vladlis** [comment]
@rnveach 
I had to reimplement original PR and your comments are not actual now, so I did not answer to each.

I have generated some reports with LAMBDA:
https://vladlis.github.io/rcurly/lambda/same/checkstyle/site/
https://vladlis.github.io/rcurly/lambda/same/infinispan/site/
https://vladlis.github.io/rcurly/lambda/same/jOOL/site/
https://vladlis.github.io/rcurly/lambda/same/openjdk8/site/
https://vladlis.github.io/rcurly/lambda/same/protonpack/site/
https://vladlis.github.io/rcurly/lambda/same/RxJava/site/

https://vladlis.github.io/rcurly/lambda/alone/checkstyle/site/
https://vladlis.github.io/rcurly/lambda/alone/infinispan/site/
https://vladlis.github.io/rcurly/lambda/alone/jOOL/site/
https://vladlis.github.io/rcurly/lambda/alone/openjdk8/site/
https://vladlis.github.io/rcurly/lambda/alone/protonpack/site/
https://vladlis.github.io/rcurly/lambda/alone/RxJava/site/

https://vladlis.github.io/rcurly/lambda/alone_or_singleline/checkstyle/site/
https://vladlis.github.io/rcurly/lambda/alone_or_singleline/infinispan/site/
https://vladlis.github.io/rcurly/lambda/alone_or_singleline/jOOL/site/
https://vladlis.github.io/rcurly/lambda/alone_or_singleline/openjdk8/site/
https://vladlis.github.io/rcurly/lambda/alone_or_singleline/protonpack/site/
https://vladlis.github.io/rcurly/lambda/alone_or_singleline/RxJava/site/

I didn't review all yet, but already see some controversial violations.
@rnveach @romani @MEZk 
Please take a look at reports and Inputs and share your opinion. Especially with `same` option

**Comment by Vladlis** [comment]
Is this should be a violation for SAME option?
https://vladlis.github.io/rcurly/lambda/same/jOOL/site/xref/jOOL/src/test/java/org/jooq/lambda/CheckedBiConsumerTest.html#L50

**Comment by romani** [comment]
@Vladlis , please generate diff reports. By diff reports you could fix only your cases and left other cases to next issue. This Check especially in "same" mode is very controversial. Usage of same on tokens that could not be multi-blocks (Lambda is not multi-block) is misconfiguration problem.

**Comment by Vladlis** [comment]
@romani
Then what mode should we use for google_checks or checkstyle_checks?

**Comment by romani** [comment]
@Vladlis , please create separate issue for google_check and checkstyle_checks (I presume we should use `alone_or_singleline`).

**Comment by Vladlis** [comment]
@romani 
Issues are opened: https://github.com/checkstyle/checkstyle/issues/4177 and https://github.com/checkstyle/checkstyle/issues/4178

Diff reports with default tokens show no regression:
https://vladlis.github.io/rcurly/diff/alone/
https://vladlis.github.io/rcurly/diff/same/
https://vladlis.github.io/rcurly/diff/alone_or_singleline/

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 4]
```diff
@@ -325,6 +355,7 @@ private static boolean isAnonInnerClassInit(DetailAST lcurly) {
     // -@cs[JavaNCSS] getDetails() method is a huge SWITCH, it has to be monolithic
     // -@cs[ExecutableStatementCount] getDetails() method is a huge SWITCH, it has to be monolithic
     // -@cs[NPathComplexity] getDetails() method is a huge SWITCH, it has to be monolithic
+    // -@cs[CyclomaticComplexity] getDetails() method is a huge SWITCH, it has to be monolithic
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: What about such suppression ?
```
-@cs[JavaNCSS,ExecutableStatementCount,*Complexity] method is a huge SWITCH, it has to be monolithic
```
if description is too long - we could wrap it on second line.

File: config/checkstyle_checks.xml [hunk 5]
```diff
@@ -47,7 +47,7 @@
   <module name="SuppressWithNearbyCommentFilter">
     <property name="commentFormat" value="-@cs\[(\w{8,})\] \w[\(\)\-\.\'\`\,\:\;\w ]{10,}"/>
     <property name="checkFormat" value="$1"/>
-    <property name="influenceFormat" value="3"/>
+    <property name="influenceFormat" value="4"/>
```
Comment (romani) [inline config/checkstyle_checks.xml]: I am not ok, see my proposal below

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 6]
```diff
@@ -141,6 +143,7 @@ public void setShouldStartLine(boolean flag) {
             TokenTypes.LITERAL_FINALLY,
             TokenTypes.LITERAL_IF,
             TokenTypes.LITERAL_ELSE,
+            TokenTypes.LAMBDA,
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: @romani Not all acceptable tokens in this check are default.
It is ok to have `lambda` here?
I believe you said we add default tokens when token won't do much damage. How did we decide this for lambda in this check?

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java [hunk 7]
```diff
@@ -271,6 +275,19 @@ private static boolean shouldBeAloneOnLineWithAloneOrSinglelineOption(Details de
...
+     * Whether last right curly should be alone on line.
+     * @param details details for validation.
+     * @return true, if last right curly should be alone on line.
+     */
+    private static boolean shouldLastRcurlyBeAloneOnLine(Details details) {
+        final boolean isTokenLambda = details.lcurly.getParent().getType() == TokenTypes.LAMBDA;
+        return !isTokenLambda
+                    && details.rcurly.getLineNo() == details.nextToken.getLineNo()
+                || isTokenLambda
+                    && getNextToken(details.nextToken).getLineNo() == details.rcurly.getLineNo();
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Why aren't these last 2 lines the same as `isAloneOnLineLambda` implementation?
Only difference is one uses `==` and the other uses `!=`.
Both methods talk about being alone on line. Shouldn't this one call the other?

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSinglelineLambda.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        Stream.of("Hello").filter(s -> {
+                return s != null;
+            }
+        ).collect(Collectors.toList());
+
+        Stream.of("Hello").filter(s -> {
+                return s != null;
+        }).collect(Collectors.toList());    // violation
+
+        Stream.of("Hello").filter(s -> {return s != null;}).collect(Collectors.toList());   // violation
```
Comment (rnveach) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSinglelineLambda.java]: Why is this a violation? It is a singleline.

**Comment by rnveach** [comment]
TC has one error which seems valid.
> RightCurlyCheck.java (1)
364: getDetails() switch has too low of a branch density (20%)

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 6 above)
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Good catch, as are not ready to provide users good support for lambda (anonymous classes) . It should not be a default.

Reference to #4177 and #4178

**Comment by romani** [comment]
@Vladlis , ping.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSinglelineLambda.java (hunk 8 above)
Comment (Vladlis) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/blocks/InputRightCurlyAloneOrSinglelineLambda.java]: This is how check behaves currently, check out this [example](http://rveach.no-ip.org/checkstyle/checkstyle.php?action=view&config=230681bd222b19e3f97ffdc5c4fd002c&code=34a35d325d4bf900a839f3b4adad317d&checkstyle=checkstyle-7.6.1-all.jar&printTree=).

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 4 above)
Comment (Vladlis) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Done

File: config/checkstyle_checks.xml (hunk 5 above)
Comment (Vladlis) [inline config/checkstyle_checks.xml]: Done

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 6 above)
Comment (Vladlis) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Fixed

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java (hunk 7 above)
Comment (Vladlis) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/blocks/RightCurlyCheck.java]: Fixed

**Comment by Vladlis** [comment]
TC violation is suppressed. Looks like refactoring will be done in https://github.com/checkstyle/checkstyle/pull/4173

**Pull request merged**